### PR TITLE
FlyThrough JPG Export

### PR DESCRIPTION
**Pull request by Jossos**
- Implemented Export option for JPG sequence
- Fixes [AB#2145](https://dev.azure.com/euclideon/57c3bd2a-8f94-4578-b195-2e4fa9d8f295/_workitems/edit/2145)

### CODE REVIEW

File: src/rendering/vcImageRenderer.h [hunk 1]
```diff
@@ -57,4 +65,6 @@ udResult vcImageRenderer_Destroy();
 
 bool vcImageRenderer_Render(vcImageRenderInfo *pImageInfo, float encodedObjectId, const udDouble4x4 &viewProjectionMatrix, const udUInt2 &screenSize, double zScale, vcTexture *pTextureOverride = nullptr);
 
+const char *GetImageSaveFormatExtension(const vcImageSaveFormat &imageSaveFormat);
```
Comment (samsinsane) [inline src/rendering/vcImageRenderer.h:68]: Why not do something similar to `vcISToWorldSize` above? Also, you haven't scoped this function.

File: vcGL/src/vcTexture.h [hunk 2]
```diff
@@ -71,7 +71,7 @@ udResult vcTexture_AsyncCreateFromMemory(vcTexture **ppTexture, udWorkerPool *pP
 void vcTexture_Destroy(vcTexture **ppTexture);
 
 udResult vcTexture_UploadPixels(vcTexture *pTexture, const void *pPixels, int width, int height, int depth = 1);
-udResult vcTexture_SaveImage(vcTexture *pTexture, vcFramebuffer *pFramebuffer, const char *pFilename);
+udResult vcTexture_SaveImage(vcTexture *pTexture, vcFramebuffer *pFramebuffer, const char *pFilename, const char *pExtension);
```
Comment (samsinsane) [inline vcGL/src/vcTexture.h:74]: Use `udStrEndsWithi`, or use an enum to dictate the format regardless of the extension.

File: vcGL/src/vcTextureHelper.cpp [hunk 3]
```diff
@@ -343,21 +343,36 @@ udResult vcTexture_SaveImage(vcTexture *pTexture, vcFramebuffer *pFramebuffer, c
...
+    int outLen = 0;
+
+    pWriteData = stbi_write_png_to_mem(pPixelsOut, 0, currSize.x, currSize.y, 4, &outLen);
+
+    UD_ERROR_NULL(pWriteData, udR_InternalError);
+    UD_ERROR_CHECK(udFile_Save(pOutFilepath, pWriteData, outLen));
+  }
+  else if (udStrcmp(pExtension, "jpg") == 0)
+  {
+    int jpgWriteResult = stbi_write_jpg(pOutFilepath, currSize.x, currSize.y, 4, pPixelsOut, 90);
```
Comment (samsinsane) [inline vcGL/src/vcTextureHelper.cpp:361]: This will create an inconsistency between JPG and PNG as `udFile_Save` can be extended to save to more than just a file. You should change both code paths to use the `stbi_write_<ext>_to_func` functions, I'm not sure how they work, so you may not be able to use `udFile_Save` but instead the individual `udFile` functions.

File: vcGL/src/vcTextureHelper.cpp [hunk 4]
```diff
@@ -343,21 +343,36 @@ udResult vcTexture_SaveImage(vcTexture *pTexture, vcFramebuffer *pFramebuffer, c
...
-  UD_ERROR_CHECK(udFile_Save(pFilename, pWriteData, outLen));
-  
   result = udR_Success;
 epilogue:
   if (pWriteData != nullptr)
     STBIW_FREE(pWriteData);
 
-  udFree(pFrameBufferPixels);
   udFree(pPixelsOut);
+  udFree(pFrameBufferPixels);
```
Comment (samsinsane) [inline vcGL/src/vcTextureHelper.cpp:375]: This was an unnecessary change.

File: src/scene/vcFlythrough.cpp [hunk 5]
```diff
@@ -254,6 +254,17 @@ void vcFlythrough::HandleSceneEmbeddedUI(vcState *pProgramState)
 
   if (m_state != vcFTS_Exporting)
   {
+    if (ImGui::BeginCombo(vcString::Get("flythroughExportFormat"), GetImageSaveFormatExtension(pProgramState->screenshot.imageSaveFormat)))
```
Comment (samsinsane) [inline src/scene/vcFlythrough.cpp:257]: This key doesn't seem to be in the translation file?

**Review by samsinsane — CHANGES_REQUESTED** [review]

File: src/scene/vcFlythrough.cpp (hunk 5 above)
Comment (Jossos) [reply src/scene/vcFlythrough.cpp:257]: A lot of the PR's I make won't have things in the translation file. Like Pauls first FlyThrough commit. One of the bugs I gotta fix is to put everything to do with FlyThroughs in the translation file, so I'll do all that at once

File: vcGL/src/vcTextureHelper.cpp (hunk 4 above)
Comment (Jossos) [reply vcGL/src/vcTextureHelper.cpp:375]: is this freed in vcTexture_ConvertPixels?

Comment (samsinsane) [reply vcGL/src/vcTextureHelper.cpp:375]: You moved this one line down.

**Comment by peterAdams1014** [comment]
I can see what you are trying to do here - but this needs big changes before acceptance.

1. Remove the added enum `vcImageSaveFormat`. Instead just check the extension from `udStrEndsWithi()`.
2. Inside `vcTexture_SaveImage()` you have two very different save paths. Either make none use `udFile_Save()`, or make both of them use it.

File: src/main.cpp [hunk 6]
```diff
@@ -1005,6 +1005,11 @@ int main(int argc, char **args)
   programState.activeViewportIndex = 0;
   programState.pActiveViewport = &programState.pViewports[programState.activeViewportIndex];
 
+  programState.supportedScreenshotFormats.Init(16);
```
Comment (samsinsane) [inline src/main.cpp:1008]: Use a static array at the top of the `vcFlythrough.cpp` file, it's what we do with similar UI elements. What you have done here is incredibly unnecessary.

File: vcGL/src/vcTextureHelper.cpp [hunk 7]
```diff
@@ -319,22 +320,42 @@ udResult vcTexture_ConvertPixels(const void *pPixels, vcTextureFormat formatIn,
   return result;
 }
 
+struct vcTextureCallbackResult
+{
+  udFile *pFile;
+  const char *pFilename;
```
Comment (samsinsane) [inline vcGL/src/vcTextureHelper.cpp:326]: This isn't necessary in the callback.

File: vcGL/src/vcTextureHelper.cpp [hunk 8]
```diff
@@ -319,22 +320,42 @@ udResult vcTexture_ConvertPixels(const void *pPixels, vcTextureFormat formatIn,
...
+  udResult result = udR_Success;
+};
+
+void vcTexture_StbCallback(void *context, void *data, int size)
+{
+  vcTextureCallbackResult *pData = (vcTextureCallbackResult *)context;
+  udResult writeResult = udFile_Write(pData->pFile, data, size);
```
Comment (samsinsane) [inline vcGL/src/vcTextureHelper.cpp:333]: Write this directly into `pData->result`. Currently, if the last write fails you're returning success, which is wrong. Additionally, you don't initialise `pData->result` to anything, so if every write fails, it's undefined as to what happens.

File: vcGL/src/vcTextureHelper.cpp [hunk 9]
```diff
@@ -319,22 +320,42 @@ udResult vcTexture_ConvertPixels(const void *pPixels, vcTextureFormat formatIn,
...
-  int outLen = 0;
-  unsigned char *pWriteData = nullptr;
   vcTextureFormat format;
   size_t pixelSize;
   uint8_t *pFrameBufferPixels = nullptr;
   uint8_t *pPixelsOut = nullptr;
+  int stbiResult = 1;
+  vcTextureCallbackResult callbackResult;
+  callbackResult.pFilename = pFilename;
+  callbackResult.pFile = udAllocType(udFile, 1, udAF_Zero);
```
Comment (samsinsane) [inline vcGL/src/vcTextureHelper.cpp:352]: This isn't how you use `udFile`, `udFile_Open` allocates for you, that's why you pass in a pointer to the `udFile*` and not just the `udFile*`.

File: vcGL/src/vcTextureHelper.cpp [hunk 10]
```diff
@@ -319,22 +320,42 @@ udResult vcTexture_ConvertPixels(const void *pPixels, vcTextureFormat formatIn,
...
 
   udInt2 currSize = udInt2::zero();
   vcTexture_GetSize(pTexture, &currSize.x, &currSize.y);
 
+  UD_ERROR_CHECK(udFile_Open(&callbackResult.pFile, callbackResult.pFilename, udFOF_Create | udFOF_Write));
```
Comment (samsinsane) [inline vcGL/src/vcTextureHelper.cpp:357]: Just use `pFilename`.

File: vcGL/src/vcTextureHelper.cpp [hunk 11]
```diff
@@ -343,21 +364,26 @@ udResult vcTexture_SaveImage(vcTexture *pTexture, vcFramebuffer *pFramebuffer, c
   UD_ERROR_NULL(pFrameBufferPixels, udR_MemoryAllocationFailure);
 
   vcTexture_BeginReadPixels(pTexture, 0, 0, currSize.x, currSize.y, pFrameBufferPixels, pFramebuffer);
-
   UD_ERROR_CHECK(vcTexture_ConvertPixels(pFrameBufferPixels, format, &pPixelsOut, vcTextureFormat_RGBA8, currSize.x * currSize.y));
+  
+  if (udStrEndsWithi(pFilename, "png"))
+    stbiResult = stbi_write_png_to_func(vcTexture_StbCallback, &callbackResult, (int)currSize.x, (int)currSize.y, (int)pixelSize, pPixelsOut, 0);
```
Comment (samsinsane) [inline vcGL/src/vcTextureHelper.cpp:370]: Just do the `UD_ERROR_IF` when you call, storing the `stbiResult` is unnecessary. Also, `pixelSize` is wrong, it needs to be `4`, that's why it was `4`.

File: vcGL/src/vcTextureHelper.cpp [hunk 12]
```diff
@@ -343,21 +364,26 @@ udResult vcTexture_SaveImage(vcTexture *pTexture, vcFramebuffer *pFramebuffer, c
...
+  else if (udStrEndsWithi(pFilename, "jpg"))
+    stbiResult = stbi_write_jpg_to_func(vcTexture_StbCallback, &callbackResult, (int)currSize.x, (int)currSize.y, (int)pixelSize, pPixelsOut, 0);
+  else
+    UD_ERROR_SET(udR_InvalidParameter_);
```
Comment (samsinsane) [inline vcGL/src/vcTextureHelper.cpp:374]: `udR_Unsupported` would make more sense.

File: vcGL/src/vcTextureHelper.cpp [hunk 13]
```diff
@@ -343,21 +364,26 @@ udResult vcTexture_SaveImage(vcTexture *pTexture, vcFramebuffer *pFramebuffer, c
   UD_ERROR_NULL(pFrameBufferPixels, udR_MemoryAllocationFailure);
 
   vcTexture_BeginReadPixels(pTexture, 0, 0, currSize.x, currSize.y, pFrameBufferPixels, pFramebuffer);
-
   UD_ERROR_CHECK(vcTexture_ConvertPixels(pFrameBufferPixels, format, &pPixelsOut, vcTextureFormat_RGBA8, currSize.x * currSize.y));
+  
+  if (udStrEndsWithi(pFilename, "png"))
+    stbiResult = stbi_write_png_to_func(vcTexture_StbCallback, &callbackResult, (int)currSize.x, (int)currSize.y, (int)pixelSize, pPixelsOut, 0);
+  else if (udStrEndsWithi(pFilename, "jpg"))
+    stbiResult = stbi_write_jpg_to_func(vcTexture_StbCallback, &callbackResult, (int)currSize.x, (int)currSize.y, (int)pixelSize, pPixelsOut, 0);
```
Comment (samsinsane) [inline vcGL/src/vcTextureHelper.cpp:372]: Same as above.

File: vcGL/src/vcTextureHelper.cpp [hunk 14]
```diff
@@ -343,21 +364,26 @@ udResult vcTexture_SaveImage(vcTexture *pTexture, vcFramebuffer *pFramebuffer, c
...
   result = udR_Success;
 epilogue:
-  if (pWriteData != nullptr)
-    STBIW_FREE(pWriteData);
+
+  UD_ERROR_CHECK(udFile_Close(&callbackResult.pFile));
 
   udFree(pFrameBufferPixels);
   udFree(pPixelsOut);
+  udFree(callbackResult.pFile);
```
Comment (samsinsane) [inline vcGL/src/vcTextureHelper.cpp:386]: As mentioned above, this is unnecessary as it's done for you by `udFile_Close`.

**Review by samsinsane — CHANGES_REQUESTED** [review]

File: src/main.cpp [hunk 15]
```diff
@@ -1004,7 +1004,7 @@ int main(int argc, char **args)
   programState.pViewports = udAllocType(vcViewport, vcMaxViewportCount, udAF_Zero);
   programState.activeViewportIndex = 0;
   programState.pActiveViewport = &programState.pViewports[programState.activeViewportIndex];
-
+
```
Comment (samsinsane) [inline src/main.cpp:1007]: You've left whitespace on this line, revert this.

File: src/main.cpp [hunk 16]
```diff
@@ -1174,7 +1174,7 @@ int main(int argc, char **args)
 
   vcProject_Deinit(&programState, &programState.activeProject);
   vcTexture_Destroy(&programState.image.pImage);
-
+
```
Comment (samsinsane) [inline src/main.cpp:1177]: You've left whitespace on this line, revert this.

File: vcGL/src/vcTextureHelper.cpp [hunk 17]
```diff
@@ -319,22 +320,35 @@ udResult vcTexture_ConvertPixels(const void *pPixels, vcTextureFormat formatIn,
   return result;
 }
 
+struct vcTextureCallbackResult
+{
+  udFile *pFile;
+  udResult result = udR_Success;
```
Comment (samsinsane) [inline vcGL/src/vcTextureHelper.cpp:326]: Don't initialise this here, do it below when you declare `callbackResult`.

File: vcGL/src/vcTextureHelper.cpp [hunk 18]
```diff
@@ -319,22 +320,35 @@ udResult vcTexture_ConvertPixels(const void *pPixels, vcTextureFormat formatIn,
...
     return udR_InvalidParameter_;
 
   udResult result;
-  int outLen = 0;
-  unsigned char *pWriteData = nullptr;
   vcTextureFormat format;
   size_t pixelSize;
   uint8_t *pFrameBufferPixels = nullptr;
   uint8_t *pPixelsOut = nullptr;
+  vcTextureCallbackResult callbackResult;
```
Comment (samsinsane) [inline vcGL/src/vcTextureHelper.cpp:345]: `= { nullptr, udR_Success };`

**Review by samsinsane — CHANGES_REQUESTED** [review]

File: vcGL/src/vcTextureHelper.cpp [hunk 19]
```diff
@@ -343,18 +357,21 @@ udResult vcTexture_SaveImage(vcTexture *pTexture, vcFramebuffer *pFramebuffer, c
...
+  UD_ERROR_CHECK(callbackResult.result);
 
-  UD_ERROR_CHECK(udFile_Save(pFilename, pWriteData, outLen));
-  
   result = udR_Success;
 epilogue:
-  if (pWriteData != nullptr)
-    STBIW_FREE(pWriteData);
+
+  UD_ERROR_CHECK(udFile_Close(&callbackResult.pFile));
```
Comment (samsinsane) [inline vcGL/src/vcTextureHelper.cpp:374]: This has the ability to create an infinite loop, and it has the ability to make this function appear as if it succeeded when it didn't. In general, don't use these macros or set `result` after the `epilogue`.

**Review by samsinsane — CHANGES_REQUESTED** [review]

File: src/main.cpp [hunk 20]
```diff
@@ -1186,7 +1186,6 @@ int main(int argc, char **args)
 
   vcProject_Deinit(&programState, &programState.activeProject);
   vcTexture_Destroy(&programState.image.pImage);
-  vcTextureCache_DeinitGlobal();
```
Comment (samsinsane) [inline src/main.cpp:1189]: You need to pay attention to the changes that you're making.

**Review by samsinsane — CHANGES_REQUESTED** [review]

**Review by samsinsane — APPROVED** [review]

**Review by samsinsane — CHANGES_REQUESTED** [review]
This has merge conflicts again.

**Review by samsinsane — DISMISSED** [review]

**Review by samsinsane — APPROVED** [review]

**Review by samsinsane — APPROVED** [review]
